lib/list_versions.go

```diff
@@ -1,172 +1,194 @@
-		// without the ending '"' pre-release folders would be tried and break.
-		semver = `\/(\d+\.\d+\.\d+)\/?"`
-	}
-	r, _ := regexp.Compile(semver)
-	for i := range result {
-		if r.MatchString(result[i]) {
-			str := r.FindString(result[i])
-			trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-			tfVersionList.tflist = append(tfVersionList.tflist, trimstr)
+	for i := range releases {
+		if !releases[i].IsPrerelease {
+			return releases[i], nil
 		}
 	}
+	return nil, nil
+}
 
-	if len(tfVersionList.tflist) == 0 {
-		fmt.Printf("Cannot get list from mirror: %s\n", mirrorURL)
+//GetTFLatestImplicit :  Get the latest implicit terraform version given the hashicorp url
+func GetTFLatestImplicit(mirrorURL string, preRelease bool, version string) (*Release, error) {
+	releases, err := GetTFReleases(mirrorURL, preRelease)
+	if err != nil {
+		return nil, err
 	}
+	if preRelease {
+		semver := fmt.Sprintf(`%s{1}\.\d+\-[a-zA-z]+\d*`, version)
+		r, err := regexp.Compile(semver)
+		if err != nil {
+			return nil, err
+		}
+		for _, release := range releases {
+			if r.MatchString(release.Version) {
+				fmt.Printf("Matched version: %s\n", release.Version)
+				return release, nil
+			}
+		}
+		return nil, fmt.Errorf("Error: no match for requested version: %s", version)
+	} else {
+		version = fmt.Sprintf("~> %v", version)
+		semv, err := SemVerParser(&version, releases)
+		return semv, err
+	}
+}
+
+// httpGet : generic http get client for the given url and query parameters.
+func httpGet(url *url.URL, values url.Values) (*http.Response, error) {
+	url.RawQuery = values.Encode()
 
-	return tfVersionList.tflist, nil
+	res, err := http.Get(url.String())
+	if err != nil {
+		return nil, fmt.Errorf("[Error] : Retrieving contents from url %s\n: %q", url, err)
 
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("[Error] : non-200 response code during request: %d: Http status: %s", res.StatusCode, res.Status)
+	}
+	return res, nil
 }
 
-//GetTFLatest :  Get the latest terraform version given the hashicorp url
-func GetTFLatest(mirrorURL string) (string, error) {
+// getReleases : subfunc for GetTFReleases, used in a loop to get all terraform releases given the hashicorp url
+func getReleases(url *url.URL, values url.Values) ([]*Release, error) {
+	var releases []*Release
+	resp, errURL := httpGet(url, values)
+	if errURL != nil {
+		return nil, fmt.Errorf("[Error] : Getting url: %q", errURL)
+	}
+	defer resp.Body.Close()
 
-	result, error := GetTFURLBody(mirrorURL)
-	if error != nil {
-		return "", error
-	}
-	// Getting versions from body; should return match /X.X.X/ where X is a number
-	semver := `\/(\d+\.\d+\.\d+)\/?"`
-	r, _ := regexp.Compile(semver)
-	for i := range result {
-		if r.MatchString(result[i]) {
-			str := r.FindString(result[i])
-			trimstr := strings.Trim(str, "/\"") //remove '/' or '"' from /X.X.X/" or /X.X.X"
-			return trimstr, nil
-		}
+	if resp.StatusCode != http.StatusOK {
```

Comment:
Is this something expected? 🤔 
![image](https://user-images.githubusercontent.com/1274789/169598662-3f52d3c1-cf2e-4c2d-8e1c-669408cf45a7.png)

Reply:
Nope that was intended to be deleted, cleaning it up now!